cpu/qn908x/periph/spi.c

```diff
@@ -0,0 +1,341 @@
+     * 0 divides by 1. */
+    spi_bus->DIV = divider - 1;
+}
+
+void spi_init(spi_t bus)
+{
+    assert(bus < SPI_NUMOF);
+    const spi_conf_t *const conf = &spi_config[bus];
+    SPI_Type *const spi_bus = conf->dev;
+
+    int flexcomm_num = flexcomm_init((FLEXCOMM_Type *)spi_bus, FLEXCOMM_ID_SPI);
+    DEBUG("[spi] init: bus=%u, flexcomm=%d\n", (unsigned)bus, flexcomm_num);
+    assert(flexcomm_num >= 0);
+
+    /* Set controller mode, but don't enable it. All CS are active low. MSB
+     * first bit order (standard). */
+    spi_bus->CFG = SPI_CFG_MASTER_MASK;
+    /* Configure to use the RX and TX FIFO. */
+    spi_bus->FIFOCFG = SPI_FIFOCFG_ENABLETX_MASK | SPI_FIFOCFG_ENABLERX_MASK;
+    locks[bus] = (mutex_t)MUTEX_INIT_LOCKED;
+    spi_init_pins(bus);
+}
+
+void spi_init_pins(spi_t bus)
+{
+    assert(bus < SPI_NUMOF);
+    const spi_conf_t *const conf = &spi_config[bus];
+
+    const uint32_t mask = conf->dev == (SPI_Type *)FLEXCOMM2_BASE
+                          ? _spi_func5_mask_fc2
+                          : 0xffffffff;
+    gpio_init_mux(conf->copi_pin,
+                  ((1u << GPIO_T_PIN(conf->copi_pin)) & mask) ? 5 : 4);
+    gpio_init_mux(conf->cipo_pin,
+                  ((1u << GPIO_T_PIN(conf->cipo_pin)) & mask) ? 5 : 4);
+    gpio_init_mux(conf->clk_pin,
+                  ((1u << GPIO_T_PIN(conf->clk_pin)) & mask) ? 5 : 4);
+    /* Enables the SPI block and sets it to idle. */
+    conf->dev->CFG |= SPI_CFG_ENABLE_MASK;
+    mutex_unlock(&locks[bus]);
+}
+
+int spi_init_cs(spi_t bus, spi_cs_t cs)
+{
+    /* Initializing the CS pin doesn't require to acquire the mutex since each
+     * peripheral has its own independent CS pin. */
+    if (bus >= SPI_NUMOF) {
+        return SPI_NODEV;
+    }
+    const spi_conf_t *const conf = &spi_config[bus];
+    gpio_t pin = cs;
+
+    if (GPIO_T_IS_HWCS(cs)) {
+        /* The gpio_t value comes from the board config rather than the cs
+         * variable itself when a HWCS number is passed. */
+        pin = conf->cs_pin[GPIO_T_HWCS(cs)];
+    }
+    if (!gpio_is_valid(pin)) {
+        return SPI_NOCS;
+    }
+    DEBUG("[spi] init_cs: cs=0x%.4" PRIx16 " pin=0x%.4" PRIx16 "\n", cs, pin);
+
+    if (GPIO_T_IS_HWCS(cs)) {
+        const uint32_t mask = conf->dev == (SPI_Type *)FLEXCOMM2_BASE
+                              ? _spi_func5_mask_fc2
+                              : 0xffffffff;
+        gpio_init_mux(pin, ((1u << GPIO_T_PIN(pin)) & mask) ? 5 : 4);
+    }
+    else {
+        gpio_init(pin, GPIO_OUT);
+        gpio_set(pin);
+    }
+
+    return SPI_OK;
+}
+
+#ifdef MODULE_PERIPH_SPI_RECONFIGURE
+void spi_deinit_pins(spi_t bus)
+{
+    assert(bus < SPI_NUMOF);
+    mutex_lock(&locks[bus]);
+    const spi_conf_t *const conf = &spi_config[bus];
+
+    /* Disables the SPI block. It must be already idle. */
+    conf->dev->CFG &= ~SPI_CFG_ENABLE_MASK;
+
+    gpio_init(conf->copi_pin, GPIO_IN);
+    gpio_init(conf->cipo_pin, GPIO_IN);
+    gpio_init(conf->clk_pin, GPIO_IN);
```

Comment:
Do we need to also disable the mux on the HWCS pins?

Reply:
That's a good question, I don't know. There is a `spi_init_pins` and a `spi_init_cs` but there's no `spi_deinit_cs` and the `spi_deinit_pins` documentation doesn't say anything about CS. I guess you are free to call `gpio_init` on a CS line whenever you want? The only other CPU that implements deinit is sam0_common and that doesn't touch the CS lines.

You are also free to call gpio_init on any of these pins without de-initializing them for this particular CPU, but calling deinit doesn't hurt to make sure the mutex is held.

Reply:
I think there is also no other CPU that implements HW CS, so this didn't come up.

> You are also free to call gpio_init on any of these pins without de-initializing them for this particular CPU,

Well then no further action is needed :)